src/StoreApi/Utilities/OrderController.php

```diff
@@ -103,26 +104,45 @@ public function sync_customer_data_with_order( \WC_Order $order ) {
 	 * By this point we have an order populated with customer data and items.
 	 *
 	 * @throws RouteException Exception if invalid data is detected.
-	 *
 	 * @param \WC_Order $order Order object.
 	 */
 	public function validate_order_before_payment( \WC_Order $order ) {
-		$coupons = $order->get_coupon_codes();
+		$this->validate_coupons( $order );
+		$this->validate_email( $order );
+		$this->validate_addresses( $order );
+	}
 
+	/**
+	 * Convert a coupon code to a coupon object.
+	 *
+	 * @param string $coupon_code Coupon code.
+	 * @return \WC_Coupon Coupon object.
+	 */
+	protected function get_coupon( $coupon_code ) {
+		return new \WC_Coupon( $coupon_code );
+	}
+
+	/**
+	 * Validate coupons applied to the order and remove those that are not valid.
+	 *
+	 * @throws RouteException Exception if invalid data is detected.
+	 * @param \WC_Order $order Order object.
+	 */
+	protected function validate_coupons( \WC_Order $order ) {
 		$coupon_errors = [];
+		$coupon_codes  = $order->get_coupon_codes();
+		$coupons       = array_map( [ $this, 'get_coupon' ], $coupon_codes );
+		$validators    = [ 'validate_coupon_email_restriction', 'validate_coupon_usage_limit' ];
 
-		foreach ( $coupons as $coupon_code ) {
+		foreach ( $coupons as $coupon ) {
 			$coupon = new \WC_Coupon( $coupon_code );
 
-			try {
-				$this->validate_coupon_email_restriction( $coupon, $order );
-			} catch ( \Exception $error ) {
-				$coupon_errors[ $coupon_code ] = $error->getMessage();
-			}
-			try {
-				$this->validate_coupon_usage_limit( $coupon, $order );
-			} catch ( \Exception $error ) {
-				$coupon_errors[ $coupon_code ] = $error->getMessage();
+			foreach ( $validators as $validator ) {
```

Comment:
Might be a good place for usage of [`array_walk`](https://www.php.net/manual/en/function.array-walk.php) to keep things a bit more concise?